### PR TITLE
refactor: fuel SDK implementation

### PR DESCRIPTION
**Pull request by arthurgeron**
# Pull Request: Refactor and Optimize Fuel Integration

> [!WARNING]  
> LICENSE:
> The Fuel Support team submits this PR to help you use the fuel SDKs. It is provided "as is" and without warranty for any purpose. We are not participants in the project at hand. All the rights of this code are reserved to the authors of this repository.

## Overview

This PR addresses issues in the Fuel integration, including bugs, anti-patterns, and code optimizations. The changes focus on improving performance, maintainability, and adhering to best practices.

## Changes Implemented

	•	Updated Fuels Connectors to latest version
	•	Updated Fuel SDK to 0.95.0.
	•	Updated wagmi version to align with the connectors.
	•	Fixed Rabby wallet and other EVMs not connecting
	•	Configured pnpm to use fixed versions to ensure consistency across environments and reduce potential conflicts.
	•	Context and Provider Optimization
	•	Avoided recreating Context API provider data inline and Fuel SDK configurations during render.
	•	Prevented passing inline objects to providers’ value fields to minimize unnecessary re-renders.
	•	State Management Enhancements
	•	Created a store for market address-based contracts with fast initialization. Contract Initialization moved to a Zustand store that initializes with the provider as soon as possible. Finishing the Landing Page's loading much earlier.
	•	Updated all usages of useMarket to utilize selectors, avoiding destructuring of the Zustand store and preventing leaked renders.
	•	Fixed issues with recreating contracts during fetch calls by initializing them appropriately.
	•	Code Cleanup
	•	Removed duplicated exports.
	•	Eliminated duplicate instantiation of queries.
	•	Query Key Optimization
	•	Resolved excessive refetching caused by passing objects as query key references in React Query.
	•	Replaced the use of queries for synchronous operations with useMemo for better performance.
	•	Linting and Configuration
	•	Addressed dangerous Biome.js configurations, particularly the disabling of critical rules like the “Rules of Hooks”.
	•	Fixed the redeclaration of the Hermes client on each render to prevent performance issues.

### Rendering and Instantiation Concerns

	•	Avoid Pre-Rendering Configurations
	•	Pre-rendering configurations and objects passed to providers, especially instantiations, can lead to unnecessary computations and performance degradation.
	•	Avoid Passing Inline Objects to Providers
	•	Passing inline objects to the provider’s value prop causes new object references on every render, triggering unnecessary re-renders.

### State Management Practices

	•	Destructuring the Zustand store can lead to unintended re-renders. Using static selectors is recommended to prevent leaked renders. While shallow comparisons can work for smaller stores that don’t change frequently, selectors offer a more reliable solution.
	•	Declaring classes (e.g., Contracts) during fetch calls can cause performance bottlenecks and should be avoided.
	•	AssetsConfig and collateralConfig were derived from the same hook side by side in useMaxWithdrawableCollateral, leading to redundant computations.

### React Query Usage

	•	Passing objects as keys in React Query can cause excessive refetching because React Query depends on the order and referential equality of keys. Always pass simple or directly stringifiable objects.
	•	Using React Query for synchronous operations is inefficient. useMemo should be used instead to memoize values without the overhead of asynchronous query management.

### Linting and Configuration Issues
Dangerous Disabling of Biome.js Rules.  

	•	Disabling critical Biome.js rules like the “Rules of Hooks” is dangerous. These rules enforce the correct usage of React hooks, ensuring they are called in the same order on every render. Violating these rules can lead to runtime errors and unpredictable behavior.

### Dependency Management

	•	Added pnpm overrides for the Fuel SDK until pythnetwork and other dependencies that use the SDK are updated.

### Numeric Operations and Big Numbers
Improper Handling of Big Numbers;   

	•	Manual concatenation and summation of numbers can lead to incorrect values (e.g., displaying 1-e8 in input fields). Doing mathematical operations with decimals in JS should be avoided at all costs, it’s recommended to use Fuel’s BN library for accurate handling and formatting of big numbers.
CleanShot 2024-10-10 at 22 54 59

### Client Reinitialization

	•	Redeclaring the Hermes client on each render can cause performance issues and unexpected behavior. The client should be initialized once and reused to maintain efficiency.

### CODE REVIEW

**Comment by martines3000** [comment]
Comments regarding react query usage:
1. This statement is wrong, based on TanStack docs: https://tanstack.com/query/latest/docs/framework/react/guides/query-keys#query-keys-are-hashed-deterministically
2. While using TanStack Query for synchronous operations is less efficient than useMemo, I believe it's a reasonable trade-off for code consistency. Combining this approach with the first point eliminates the need to implement custom 'stringify' functions, as TanStack Query handles everything. Additionally, it facilitates a smoother transition to asynchronous operations if future requirements change.

**Comment by martines3000** [comment]
Can `pnpm overrides` cause problems, if for example the Pyth SDK uses old types ? (runtime errors ?)

File: apps/frontend/src/components/Providers/index.tsx [hunk 1]
```diff
@@ -90,42 +108,41 @@ const wagmiConfig = createWagmiConfig({
...
     });
-  } catch (error) {
-    console.error(error);
-  }
-}
+  }, []);
 
-export const Providers = ({ children }: { children: ReactNode }) => {
-  const queryClient = getQueryClient();
+  const fuelConfig = useMemo(
```
Comment (martines3000) [inline apps/frontend/src/components/Providers/index.tsx:130]: 1. Do we need this here ?

The configuration was copied from this example: https://github.com/FuelLabs/fuel-connectors/blob/9f18963a0a2330aeb1a8c296a28d7a441d52022b/examples/react-next/src/components/FuelProviders.tsx#L23

2. This setup also causes issues on mainnet:
image

3. Is there a problem to do it the old way and initialize 2 providers as we did before ?

**Comment by martines3000** [comment]
Comments regarding state Management Practices
1. This is a good change and we will try to avoid relying on shallow in the future as it can easily go wrong. Just a small correction, the store size has nothing to do with shallow, as our `marketStore` works fine with it. The limitations of `shallow` can be seen when using [complex structures](https://zustand.docs.pmnd.rs/apis/shallow#shallow) (objects, maps, etc..) inside the store (which we don't use). 

Can we maybe [autogenerate](https://zustand.docs.pmnd.rs/guides/auto-generating-selectors) the selectors ?

File: .changeset/giant-ravens-drop.md [hunk 2]
```diff
@@ -0,0 +1,5 @@
+---
+"swaylend-dapp": patch
```
Comment (petertonysmith94) [inline .changeset/giant-ravens-drop.md:2]:
```suggestion
"swaylend-dapp": patch
"swaylend-indexer-sentio": patch
```

File: .changeset/loud-cougars-admire.md
Comment (petertonysmith94) [inline .changeset/loud-cougars-admire.md:1]: Delete

File: apps/frontend/package.json [hunk 3]
```diff
@@ -13,7 +13,7 @@
   },
   "dependencies": {
     "@clerk/nextjs": "^5.6.4",
-    "@fuel-ts/contract": "^0.94.9",
+    "@fuel-ts/contract": "^0.95.0",
     "@fuels/connectors": "0.31.6",
     "@fuels/react": "0.31.6",
```
Comment (petertonysmith94) [inline apps/frontend/package.json:18]:
```suggestion
    "@fuels/connectors": "0.32.0",
    "@fuels/react": "0.32.0",
```

File: apps/frontend/src/components/Providers/index.tsx (hunk 1 above)
Comment (arthurgeron) [reply apps/frontend/src/components/Providers/index.tsx:130]: The error isn't related with the useMemo, everything passwed to our providers should always be either primitive, static, global or memoized

**Comment by arthurgeron** [comment]
> Comments regarding state Management Practices
> 
> 1. This is a good change and we will try to avoid relying on shallow in the future as it can easily go wrong. Just a small correction, the store size has nothing to do with shallow, as our `marketStore` works fine with it. The limitations of `shallow` can be seen when using [complex structures](https://zustand.docs.pmnd.rs/apis/shallow#shallow) (objects, maps, etc..) inside the store (which we don't use).
> 
> Can we maybe [autogenerate](https://zustand.docs.pmnd.rs/guides/auto-generating-selectors) the selectors ?



Shallow:
- Does not run when you call the store with destruct, with selectors. Meaning every store update that passes shallow will leak renders to all components accessing the store;
- Runs through every property key/value pair of every object before each update, hence the reason it's not recommended to frequently updated stores;

https://zustand.docs.pmnd.rs/apis/shallow#comparing-objects-returns-false-even-if-they-are-identical.

**Comment by arthurgeron** [comment]
> Comments regarding react query usage:
> eliminates the need to implement custom 'stringify' functions, as TanStack Query handles everything. Additionally, it facilitates a smoother transition to asynchronous operations if future requirements change.

Custom stringify functions should never be used for useMemo dependency arrays, if that's what you mean, since they'd run every render. TanStack Query will map your entire JSON objects, every render, it's far for the intended purpose. Ideally strings/ids should go in the key array, worst case scenario simple objects. But Query was never meant to be used for synchronous expressions, it's an Agnostic data-fetching library!

> it facilitates a smoother transition to asynchronous operations if future requirements change.
Imho I don't think it's worth the compromise of slowing down all your expression evaluations and stringifying JSONs on query keys for every render.

**Comment by arthurgeron** [comment]
> Can `pnpm overrides` cause problems, if for example the Pyth SDK uses old types ? (runtime errors ?)

I hope you have tested this PR thoroughly as I did before merging. Regarding the SDK, there were several crucial changes, the risk of having different decimal cases and other nasty bugs, hence the override.
Old types won't be a problem since they weren't removed, but changed. Without the override type conflicts is exactly what would happen

**Comment by martines3000** [comment]
I don't know how you "tested" it, but it completely broke the Markets page behaviour and also the modals on the dashboard 😓. I fixed everything in #173.

Regarding Zustand, selectors are the way to go, I agree. More performant, less room for errors. Thanks for showing this, will definitely avoid shallow from now on 🔥.

Regarding other suggestions, we will keep them in mind 👍

**Comment by petertonysmith94** [comment]
@arthurgeron and @martines3000 great work peeps - we're all on the same team.

**Pull request merged**